stm32 comparator

KozhinovAlexander opened the pull request:
Add stm32 comparator driver and simple comparator sample.

I've used to test it on nucleo_h745zi_q (m7 core only). 

Build command with newly added sample:

```bash
west build -p auto -d tests/drivers/comparator/gpio_loopback/build -b nucleo_h745zi_q/stm32h745xx/m7 tests/drivers/comparator/gpio_loopback  
```

My build result output:

```bash
-- Zephyr version: 4.0.99 (...../zephyr), build: v4.0.0-4417-g2c7031229fe0
[153/153] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       31316 B         1 MB      2.99%
             RAM:        6848 B       512 KB      1.31%
          EXTMEM:          0 GB       256 MB      0.00%
           SRAM1:          0 GB       128 KB      0.00%
           SRAM2:          0 GB       128 KB      0.00%
           SRAM3:         16 KB        32 KB     50.00%
           SRAM4:          0 GB        64 KB      0.00%
            DTCM:          0 GB       128 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

The resulting output in the console could look like:

```bash
*** Booting Zephyr OS build v4.0.0-4417-g963f50d621cd ***
Running TESTSUITE comparator_gpio_loopback
===================================================================
START - test_get_output
 PASS - test_get_output in 0.005 seconds
===================================================================
START - test_no_trigger_no_pending
 PASS - test_no_trigger_no_pending in 0.003 seconds
===================================================================
START - test_trigger_both_edges_pending
 PASS - test_trigger_both_edges_pending in 0.004 seconds
===================================================================
START - test_trigger_callback
 PASS - test_trigger_callback in 0.004 seconds
===================================================================
START - test_trigger_falling_edge_pending
 PASS - test_trigger_falling_edge_pending in 0.004 seconds
===================================================================
START - test_trigger_rising_edge_pending
 PASS - test_trigger_rising_edge_pending in 0.004 seconds
===================================================================
TESTSUITE comparator_gpio_loopback succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [comparator_gpio_loopback]: pass = 6, fail = 0, skip = 0, total = 6 duration = 0.024 seconds
 - PASS - [comparator_gpio_loopback.test_get_output] duration = 0.005 seconds
 - PASS - [comparator_gpio_loopback.test_no_trigger_no_pending] duration = 0.003 seconds
 - PASS - [comparator_gpio_loopback.test_trigger_both_edges_pending] duration = 0.004 seconds
 - PASS - [comparator_gpio_loopback.test_trigger_callback] duration = 0.004 seconds
 - PASS - [comparator_gpio_loopback.test_trigger_falling_edge_pending] duration = 0.004 seconds
 - PASS - [comparator_gpio_loopback.test_trigger_rising_edge_pending] duration = 0.004 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

I've not tested the driver with any other  stm32, but I am planning to add support for stm32g4 soon. 
Any testers with appropriate board (any stm32h7 would be fine as a good first choice) are very welcome.

KozhinovAlexander committed:
drivers: comparator: stm32

Add comparator driver for ST's stm32 series.

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

jilaypandya commented on drivers/comparator/CMakeLists.txt:
```diff
@@ -9,5 +9,6 @@ zephyr_library_sources_ifdef(CONFIG_USERSPACE comparator_handlers.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_FAKE_COMP comparator_fake_comp.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_MCUX_ACMP comparator_mcux_acmp.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_NRF_COMP comparator_nrf_comp.c)
+zephyr_library_sources_ifdef(CONFIG_COMPARATOR_STM32_COMP comparator_stm32_comp.c)
```
sort alphabetically

KozhinovAlexander replied:
done

jilaypandya commented on drivers/comparator/Kconfig:
```diff
@@ -21,6 +21,7 @@ config COMPARATOR_INIT_PRIORITY
 rsource "Kconfig.fake_comp"
 rsource "Kconfig.mcux_acmp"
 rsource "Kconfig.nrf_comp"
+rsource "Kconfig.stm32_comp"
```
here as well

KozhinovAlexander replied:
done

bjarki-andreasen reviewed (changes requested):
The driver looks good :) The sample is not needed I believe. The test suite in `tests/drivers/comparator/gpio_loopback` validates the API behavior with a GPIO connected to the comp input. The build all test suite allows you to build varoius dt configurations `tests/drivers/build_all/comparator`, and lastly, given the nature of a comparator, just enabling the comparator shell in a blank application is sample enough :)

Please add the comp to the two test suites to validate and build the driver in CI

KozhinovAlexander commented:
Thank you for your review. 

> The driver looks good :) The sample is not needed I believe. The test suite in `tests/drivers/comparator/gpio_loopback` validates the API behavior with a GPIO connected to the comp input. The build all test suite allows you to build varoius dt configurations `tests/drivers/build_all/comparator`, and lastly, given the nature of a comparator, just enabling the comparator shell in a blank application is sample enough :)
> 

Makes sense. I have few questions just for my understanding:

- Is there also any HW or only twister, where tests are running?

- Will test-suite test the stm32 driver too?

> Please add the comp to the two test suites to validate and build the driver in CI 

Will do of course. Thanks.

Regarding comparator sample: I think - the major advantage of the sample is to give a very basic example coupled with physical functionality. Especially in my experience samples allow very fast peripheral testing on new hardware, device-tree problem findings with new BSP.
Additionally all the zephyr samples give west build command for an arbitrary sample. People learn it faster, than running twister, I would assume. Thus it is always faster for them to go with sample than with test.
Of course we may discuss it in a separate PR for sample. For the sake of simplicity I will drop the sample in this PR.

bjarki-andreasen commented:
> Thank you for your review.
> 
> > The driver looks good :) The sample is not needed I believe. The test suite in `tests/drivers/comparator/gpio_loopback` validates the API behavior with a GPIO connected to the comp input. The build all test suite allows you to build varoius dt configurations `tests/drivers/build_all/comparator`, and lastly, given the nature of a comparator, just enabling the comparator shell in a blank application is sample enough :)
> 
> Makes sense. I have few questions just for my understanding:
> 
>     * Is there also any HW or only twister, where tests are running?
>
>     * Will test-suite test the stm32 driver too?

Many vendors, including STM I believe, run twister tests on hardware in their own CI, zephyr only runs emulated platforms in CI. That's why we specify fixtures in the zephyr test suites, like `gpio_loopback`, to specify what hardware they can run on :) That's why the gpio_loopback comparator test was created, its widely supported by downstream boards (and simple to set up locally of course)

I don't know exactly what their `gpio_loopback` looks like, or if they test the board you use in this PR, maybe @FRASTM knows more?

> > Please add the comp to the two test suites to validate and build the driver in CI
> 
> Will do of course. Thanks.
> 
> Regarding comparator sample: I think - the major advantage of the sample is to give a very basic example coupled with physical functionality. Especially in my experience samples allow very fast peripheral testing on new hardware, device-tree problem findings with new BSP. Additionally all the zephyr samples give west build command for an arbitrary sample. People learn it faster, than running twister, I would assume. Thus it is always faster for them to go with sample than with test. Of course we may discuss it in a separate PR for sample. For the sake of simplicity I will drop the sample in this PR.

Tests don't require twister to run, they are just normal applications with a nice framework to set up cases without needing a bunch of `printk()` everywhere.

Simply `west build -p -b nucleo_h745zi_q zephyr/tests/drivers/comparator/gpio_loopback` and flash it (after you add the board overlay for it in the test suite of course). So, its just as fast as the sample, but about 90% less duplication :)

The reason I favor the shell above a "fixed" scenario like the sample in this PR, is that actually playing around with a comparator usually requires some external signal generator to generate a voltage or an edge, the GPIO case only shows that the API "works", not that the comparator is actually measuring and behaving correctly. That is quick to see using the shell to test "real" varying signals.

KozhinovAlexander committed:
tests: drivers: comparator: gpio_loopback

Add support for nucleo_h745zi_q board
Add support for nucleo_g474re board

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

KozhinovAlexander commented:
@bjarki-andreasen Thanks for pointing out to tests. I've enjoyed using them and fixed some miss-implementation. No the driver shall follow the specification of the api better. Please take a look again.

@jilaypandya I've also added changes you requested. Please take a look.

bjarki-andreasen commented on include/zephyr/drivers/comparator/stm32_comp.h:
```diff
@@ -0,0 +1,22 @@
+
+#include <zephyr/drivers/comparator.h>
+
+#include <stm32_ll_comp.h>
```
Device driver specific functions are allowed, but they need to be implemented using zephyr's device driver model and coding style. Exposing the STM32 HAL directly in a public header is not allowed.

If a user wishes to use the HAL directly, they can simply `#include <stm32_ll_comp.h>` in their own source files and use the device directly, but doing this requires the user to not use the device driver you have created :)

So, to add a device driver specific API for the stm32 comp:
```
#include <zephyr/drivers/comparator.h>

/* Just like the wrappers, the `struct device *` is used for the context, and the config passed to it is
 * written using zephyrs coding style, so no `COMP_InitTypeDef` for example. This header has to be
 * completely decoupled from any HAL.
 */

struct comp_stm32_comp_config {
        uint32_t ...
};

int comp_stm32_comp_config(const struct device *dev, const struct comp_stm32_comp_config *config);
```

The function is then implemented in the driver's source code like a "normal" function
```
int comp_stm32_comp_config(const struct device *dev, const struct comp_stm32_comp_config *config);
{
        ...
        return 0;
}
```
See the other comparator drivers for inspiration :)

KozhinovAlexander replied:
Makes sense. Thanks for pointing out. I will make necessary changes.

KozhinovAlexander replied:
I've moved stm32 has include to corresponding .c file. There is no need at the moment in st specific functionality implementation. But it was nice to know for me. Thnx.

erwango commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,24 @@
+	default y
+	depends on DT_HAS_ST_STM32_COMP_ENABLED
+	select PINCTRL
+	select USE_STM32_LL_COMP
+	select USE_STM32_LL_EXTI
+	select USE_STM32_LL_PWR if SOC_SERIES_STM32H7X
+	select USE_STM32_LL_RCC if SOC_SERIES_STM32H7X
```
Note that use of these symbols is required only if you're making use of LL source files.
They are not required if you only use headers.

KozhinovAlexander replied:
@erwango  Thank you. I've already realized it from cmake target binding in corresponding cmake file. But anyway it was a very good one note - I wanted to review this section. Truly both LL_PWR and LL_RCC were not used s.t. .c files wer built and linked. Now I dropped both settings. All the tests still passing (at least locally).

KozhinovAlexander commented:
@bjarki-andreasen Thank you for a very good review. I'me made all the changes you requested. Please take a look again.

KozhinovAlexander commented:
@erwango Thank you for a very good review and valuable input. I've followed all the requests. At the moment my tests (see west command in description) are running well on my local setup with PB0 looped back to PB2 with a physical cable. As soon as I disconnect this cable only second test pass (see output section in PR description).  Therefor I have a equations regarding possible solution:

1. Does ST has any possibility to connect pins for the test physically on your test bench?
2. (or if 1 not possible) Is there any loopback between pb0 and pb2 in device-tree possible s.t. only corresponding overlay its file within test itself could be updated?

bjarki-andreasen reviewed (changes requested):
Please remove the empty/redundant stm32_comp.h

KozhinovAlexander commented:
> Please remove the empty/redundant stm32_comp.h

Thnx. Done.

erwango commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,22 @@
+	select USE_STM32_LL_EXTI
+
+config COMPARATOR_STM32_COMP_EXTI_HAS_LINES_64_95
+	bool "ST STM32 EXTI has lines 64 to 95"
```
If this is soc specific and user has no configuration to do, please remove the prompt ("ST [...]"), so user cannot change it

KozhinovAlexander replied:
Changed to: "STM32 EXTI has lines 64 to 95"

erwango commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,22 @@
+	default y if SOC_SERIES_STM32H7X
+
+config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
+	bool "ST STM32 comparator Miller effect handling"
```
Ditto

KozhinovAlexander replied:
Changed to: "STM32 Comparator Miller effect handling"
I hope I understood you right.

erwango commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,376 @@
+#define STM32_COMP_DT_POWER_MODE(inst)                                                             \
+	CONCAT(LL_COMP_POWERMODE_, DT_INST_STRING_TOKEN(inst, st_power_mode))
+
+#if defined(SOC_SERIES_STM32H7X)
+#define STM32_COMP_DT_INST_COMP_CONFIG_INIT(inst)                                                  \
+	{                                                                                          \
+		.PowerMode = STM32_COMP_DT_POWER_MODE(inst),                                       \
+		.InputPlus = STM32_COMP_DT_INST_P_IN(inst),                                        \
+		.InputMinus = STM32_COMP_DT_INST_N_IN(inst),                                       \
+		.InputHysteresis = STM32_COMP_DT_INST_HYST_MODE(inst),                             \
+		.OutputPolarity = STM32_COMP_DT_INST_INV_OUT(inst),                                \
+		.OutputBlankingSource = STM32_COMP_DT_INST_BLANK_SEL(inst),                        \
+	}
+#else /* SOC_SERIES_STM32H7X */
+#define STM32_COMP_DT_INST_COMP_CONFIG_INIT(inst)                                                  \
+	{                                                                                          \
+		.InputPlus = STM32_COMP_DT_INST_P_IN(inst),                                        \
+		.InputMinus = STM32_COMP_DT_INST_N_IN(inst),                                       \
+		.InputHysteresis = STM32_COMP_DT_INST_HYST_MODE(inst),                             \
+		.OutputPolarity = STM32_COMP_DT_INST_INV_OUT(inst),                                \
+		.OutputBlankingSource = STM32_COMP_DT_INST_BLANK_SEL(inst),                        \
+	}
+#endif /* SOC_SERIES_STM32H7X */
```
If the SoC specific handling is caused by the h7 comp IP being different from other STM32, please introduce a `st,stm32h7-comp` binding that can be used to distinguish implementation to use.
The interest is that for the next SoC series that has an IP similar to H7, we just have to use `st,stm32h7-comp` compatible in dts description and "voilà": no impact required in the driver to support the new series.

KozhinovAlexander replied:
Added `st,stm32h7-comp` and corresponding dt combat preprocessor check.

erwango commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,376 @@
+#define STM32_COMP_DT_POWER_MODE(inst)                                                             \
+	CONCAT(LL_COMP_POWERMODE_, DT_INST_STRING_TOKEN(inst, st_power_mode))
+
+#if defined(SOC_SERIES_STM32H7X)
```
`#if defined(CONFIG_SOC_SERIES_STM32H7X)`, but see below comment

KozhinovAlexander replied:
done.

erwango commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,179 @@
+  st,blank-sel:
+    type: string
+    description: |
+      TODO
```
That would be the right time ;)

erwango commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,376 @@
+	return LL_COMP_ReadOutputLevel(cfg->comp);
+}
+
+static int stm32_comp_set_trigger(const struct device *dev, enum comparator_trigger trigger)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	struct stm32_comp_data *data = dev->data;
+	int ret = 0;
+
+	LL_EXTI_InitTypeDef exti_init_data = {0};
+
+	LL_EXTI_StructInit(&exti_init_data);
+	exti_init_data.LineCommand = ENABLE;
+	exti_init_data.Mode = LL_EXTI_MODE_IT;
+
+	if (cfg->exti_line_number < 32U) {
+		exti_init_data.Line_0_31 = cfg->exti_line;
+	} else if (cfg->exti_line_number < 64U) {
+		exti_init_data.Line_32_63 = cfg->exti_line;
+	} else {
+		exti_init_data.Line_64_95 = cfg->exti_line;
```
For the whole exti usage, did you consider using `drivers/interrupt_controller/intc_exti_stm32.c`.
I haven't dig deep into the implementation, but I'd like to avoid code duplication and potential resource usage conflicts. See stm32 gpio driver for example on how this is used.
It is possible that some adaptation would be required to exti driver if needed.

I'll need to dig deeper on that point but that would be the main blocker to me untill we're sure this is the best option.

KozhinovAlexander replied:
No. I didn't. But it makes sense to reuse of course.  I've skimmed through `drivers/interrupt_controller/intc_exti_stm32.c` and see a lot of simplifications. For example only EXTI lines 0 to 31 will be considered, since it was primarily developed for GPIO and gpio uses lines 0 to 15. 
My verdict: I would need a deep rework of this driver, but it shall be possible. 
Although I need support regarding diversity of st's exti implementations. At the moment I am only familiar with h74xx and g4xx. 
Does `drivers/interrupt_controller/intc_exti_stm32.c` also has tests implemented?

KozhinovAlexander replied:
@erwango Would you create an issue? I can then reimplement, I think. Or do you have anybody else already working on this topic?

erwango replied:
> @erwango Would you create an issue? I can then reimplement, I think. Or do you have anyone else already working on this topic?

I'll create the issue. This was not in our short term plans as you just created the need ;)

erwango commented:
> 1. Does ST has any possibility to connect pb0 to pb2 for the test physically on the test bench?

Yes, but note that the H7 board we're using on our test farm is `nucleo_h743zi`
This is done thanks the `gpio_loopback` fixup provided in the testcase.yaml, but ideally, it needs to match other `gpio_loopback` usage across test cases

KozhinovAlexander commented:
> > 1. Does ST has any possibility to connect pb0 to pb2 for the test physically on the test bench?
> 
> Yes, but note that the H7 board we're using on our test farm is `nucleo_h743zi` This is done thanks the `gpio_loopback` fixup provided in the testcase.yaml, but ideally, it needs to match other `gpio_loopback` usage across test cases

I know for sure, that h743 and h745 are pretty similiar i.e. an overlay from h745 will do for h743 too. I could add h743 overlay too. The only open question to me is then: which two pins shall be used for the gpio_loopback? Of course this pins shall be also supported by one of the comparators as P and M inputs.

KozhinovAlexander commented:
I've added nucleo_h743zi.overlay and enabled in  the test case.

KozhinovAlexander commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,186 @@
+      Miller effect is a hidden feature of stm32, which may be enabled by setting
+      bit 1 of COMP_CxCSR register.
+      for more info see:
+        https://community.st.com/t5/stm32-mcus-products/stm32g431-comparator-bug/m-p/679540#M250779
```
@erwango Miller effect handling is so far g4 specific. Here you see the link to st's forum discussion in the parameter description. Please take a look on it.

mathieuchopstm replied:
Please create a new `stm32g4-comp` binding where this property is added if this is a series-specific feature.

KozhinovAlexander commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,20 @@
+if COMPARATOR_STM32_COMP
+
+config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
+	bool "STM32 Comparator Miller effect handling"
+	depends on COMPARATOR_STM32_COMP
+	default n
+	default y if SOC_SERIES_STM32G4X
```
It is g4 specific. I've mentioned you in in .yaml file comment for this topic.

KozhinovAlexander commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,376 @@
+
+	if (cfg->exti_line_number < 32U) {
+		exti_init_data.Line_0_31 = cfg->exti_line;
+	} else if (cfg->exti_line_number < 64U) {
+		exti_init_data.Line_32_63 = cfg->exti_line;
+	} else {
+		exti_init_data.Line_64_95 = cfg->exti_line;
```
I will check if it builds for other series like g4 today too.

KozhinovAlexander commented on dts/bindings/comparator/st,stm32-comp.yaml:
added

KozhinovAlexander commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,186 @@
+    description: |
+      Number of the Extended Interrupts and Event Controller (EXTI) interrupt
+      line connected to the COMPx event.
+      Not required, since COMPx interrupt could be routed directly to Nested
+      Vectored Interrupt Controller (NVIC).
```
> This requires some clarification. Looking the driver implementation, it looks like it is expected that `st,exti-line` is provided. So in any case, this should be fixed some way or the other.
> 
> If this use of NVIC vs EXTI depends on SoC, this should be made explicit by the compatible and provide as many compatible as it is needed to describe the various COMP IPs that the driver intends to support.

From stm32h7 refman: RM0399 Rev 4 page 1124, 29.5 COMP Interrupts:

```
There are two ways to use the comparator as interrupt source.
The comparator outputs are internally connected to the Extended interrupt and event
controller. Each comparator has its own EXTI line and can generate either interrupts or
events to make the device exit low-power modes.
The comparators also provide an interrupt line to the NVIC of CPU. This functionality is 
```

From stm32g4 refman: RM0440 Rev 8 page 770, 24.5 COMP Interrupts:

```
The comparator outputs are internally connected to the Extended interrupts and events
controller. Each comparator has its own EXTI line and can generate either interrupts or
events. The same mechanism is used to exit Sleep and Stop low-power modes.
Refer to Interrupt and events section for more details.
```

I understand it like this:

 - stm32h7 can do IRQ either over EXTI to NVIC or directly to NVIC
 - stm32g4 can do IRQ only over EXTI to NVIC

It would be interesting to know, if there are other combinations possible.

erwango replied:
> UPD: I would suggest going with EXTI only support is a first good approach. NVIC support shall be optional, but feasible.

Ok, then let's fully remove IRQ support for now as a first approach.

KozhinovAlexander replied:
> Ok, then let's fully remove IRQ support for now as a first approach.

I've marked `st,exti-line` property as required and removed sentence about NVIC. The implementation itself already follows EXTI IRQ approach:

From RM0440 Rev 8 page 770 of stm32g4 series:
```
To enable the COMPx interrupt, it is required to follow this sequence:
1. Configure and enable the EXTI line corresponding to the COMPx output event in
interrupt mode and select sensitivity to rising edge, falling edge or to both edges
2. Configure and enable the NVIC IRQ channel mapped to the corresponding EXTI lines
3. Enable COMPx
```

From RM0399 Rev 4 page 1124 of stm32h7 series:
```
Sequence to enable the COMPx interrupt through EXTI block:
1. Configure the EXTI line, receiving the comp_wkup signal, in interrupt mode, select the
rising, falling or either-edge sensitivity and enable the EXTI line.
2. Configure and enable the NVIC IRQ channel mapped to the corresponding EXTI lines.
3. Enable the COMPx.
```

The EXTI enabling functionality according to steps above can be found in the `stm32_comp_set_trigger()` function of `drivers/comparator/comparator_stm32_comp.c` file.

KozhinovAlexander commented:
@erwango So far I've worked over all your change-requests and added g4 support to the test case too.  The code line `exti_init_data.Line_64_95 = cfg->exti_line;` was a problem for g4, but I've layered it with `DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_comp)`. Now it builds.

erwango commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,20 @@
+
+config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
+	bool "STM32 Comparator Miller effect handling"
+	depends on COMPARATOR_STM32_COMP
```
Remove this line (superfluous as symbol is already under if `COMPARATOR_STM32_COMP`).

KozhinovAlexander replied:
done

erwango commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,20 @@
+if COMPARATOR_STM32_COMP
+
+config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
+	bool "STM32 Comparator Miller effect handling"
```
```suggestion
	bool
```

Providing a prompt ("[...]") makes the symbol user cofigurable, which we don't want because it only depends on SoC.

KozhinovAlexander replied:
done

KozhinovAlexander commented:
@erwango The build of tests is running all except your approved. From my side I see implementation closed (except discussed EXTI integration, which is the other story). Please take a look and approve/request changes. As soon as you approve I'll request @bjarki-andreasen and @jilaypandya to take a look aggain.

Also everyone is welcome to test it on HW you have ;)

KozhinovAlexander commented on tests/drivers/comparator/gpio_loopback/testcase.yaml:
```diff
@@ -29,3 +29,10 @@ tests:
       - nrf5340dk/nrf5340/cpuapp
+  drivers.comparator.gpio_loopback.stm32_comp:
+    extra_args:
+      - SNIPPET="gpio_loopback_stm32_comp"
```
@JarmouniA Thank you. You're absolutely right. After rebase to the newest `origin/master` I've realised that `nucleo_h745zi_q.overlay` should be renamed to `nucleo_h745zi_q_stm32h745xx_m7.overlay` indeed.
I've also dropped `extra_args` entry from `drivers.comparator.gpio_loopback.stm32_comp`.
I've putted all the changes to a separate commit 82c98b8ef0bc0bd75ad086f58ac830215db32f7d to be able to easily revert in case it will not work for all.
I also updated build command (see PR description).
Good catch :)

KozhinovAlexander replied:
@JarmouniA Regarding `SNIPPETS` extra arg: I've just followed already available test cases (e.g. copy-paste). Thus I do not see myself responsible for managing it :) If all reviewers agree, that `SNIPPETS` extra arg may be dropped I am ok with it since it was not the intention of this PR.

KozhinovAlexander replied:
> Is this needed!? the board overlays are applied automatically. `nucleo_h745zi_q.overlay` should be renamed to `nucleo_h745zi_q_stm32h745xx_m7.overlay`

Got managed it today. Snippets are note needed for comparator. They may be added, but I do not see any necessary need. Thanks, you've found my copy-paste dumb approach :), I've followed. Your sugggestio will simplify this PR.

bjarki-andreasen replied:
The reason snippets are used for the `nrf_comp` and `nrf_lpcomp` is that its two different devicetree configurations for the same board, so just using a board overlay was not feasible :) Hence, there is a "base" overlay for the board, and either snippet is included based on which configuration to test for :)

KozhinovAlexander replied:
I see we all agree. Thnx. I will stash then [d81d963](https://github.com/zephyrproject-rtos/zephyr/pull/85022/commits/d81d963d09f724aebd57f61fb324c93bd3df8784) and [cd9684a](https://github.com/zephyrproject-rtos/zephyr/pull/85022/commits/cd9684aefe1f57016cac493972c880dbcddfa365) commits.

KozhinovAlexander commented:
I've rebase to 4.1.0 and added comment to h745 overlay in the test with D27 and D33 shortage notice to make test pass, since connection of this pins is obligatory for the test.

KozhinovAlexander commented:
I would say - this one PR could be reworked after #85508 were merged.

KozhinovAlexander commented:
second rebase over latest main to ease merge

mathieuchopstm commented on drivers/comparator/Kconfig.stm32_comp:
```diff
@@ -0,0 +1,19 @@
+
+if COMPARATOR_STM32_COMP
+
+config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
+	bool
+	default n
+	default y if SOC_SERIES_STM32G4X
```
If this feature is only present on STM32G4, the following should be used instead:
```suggestion
config COMPARATOR_STM32_COMP_MILLER_EFFECT_HANDLING
	bool
	depends on SOC_SERIES_STM32G4X
	default y
```

mathieuchopstm replied:
Description should be added too. Also, since this appears to not be an officially supported feature, that aspect should be clearly documented too.

mathieuchopstm commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -1083,6 +1083,34 @@
+			reg = <0x58003810 0x200>;
+			interrupts = <137 0>;
+			st,exti-line = <20>;
+			st,blank-sel = "NONE";
```
Don't repeat default value in DTSI.
```suggestion
```

mathieuchopstm commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,185 @@
+  positive-input:
+    required: true
+    type: string
+    description: |
+      Selects positive input pin given by pinmux controller. This input can be
+      typically assigned to one of the predefinded package pins.
+      IO0 will correspond to first pinmux pin option and IO1 to the second one.
+    enum:
+      - IN0
+      - IN1
+    default: IN0
```
Description and actual values differ.
IMO, we should stick to the LL naming convention and use `IO1`/`IO2`/etc.

mathieuchopstm commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,185 @@
+      - DAC3_CH2
+      - DAC4_CH1
+      - DAC4_CH2
+      - IN0
+      - IN1
```
Same remark here: use LL convention of `IO1`/etc

mathieuchopstm commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,185 @@
+  st,power-mode:
+    type: string
+    description: |
+      Power mode is not supported by all stm32 series.
```
Nit: then it should only exist in the compatible of the series that support it.

mathieuchopstm commented on dts/bindings/comparator/st,stm32-comp.yaml:
```diff
@@ -0,0 +1,185 @@
+    description: |
+      Selects positive input pin given by pinmux controller. This input can be
+      typically assigned to one of the predefinded package pins.
+      IO0 will correspond to first pinmux pin option and IO1 to the second one.
```
From my understanding of the HW, this is misleading - I don't see what prevents assigning only one pinmux to COMP (nor how the order matters to this property).

mathieuchopstm commented on tests/drivers/comparator/gpio_loopback/boards/nucleo_h745zi_q_stm32h745xx_m7.overlay:
```diff
@@ -0,0 +1,51 @@
+};
+
+&pinctrl {
+	/omit-if-no-ref/ comp1_in_p_pb0: comp1_in_p_pb0 {
+		pinmux = <STM32_PINMUX('B', 0, ANALOG)>;
+	};
```
Unused?

mathieuchopstm commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -1083,6 +1083,34 @@
+			reg = <0x5800380c 0x200>;
+			interrupts = <137 0>;
+			st,exti-line = <20>;
+			st,blank-sel = "NONE";
```
Don't repeat default value in DTSI.
```suggestion
```

mathieuchopstm commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -662,6 +663,32 @@
+			reg = <0x40010204 0x4>;
+			interrupts = <64 0>;
+			st,exti-line = <22>;
+			st,blank-sel = "NONE";
```
Don't repeat default value in DTSI.
```suggestion
```

mathieuchopstm commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -662,6 +663,32 @@
+			reg = <0x40010200 0x4>;
+			interrupts = <64 0>;
+			st,exti-line = <21>;
+			st,blank-sel = "NONE";
```
Don't repeat default value in DTSI.
```suggestion
```

mathieuchopstm commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,378 @@
+
+#define DT_DRV_COMPAT st_stm32_comp
+
+#define LL_COMP_INPUT_PLUS_IN0 LL_COMP_INPUT_PLUS_IO1
+#define LL_COMP_INPUT_PLUS_IN1 LL_COMP_INPUT_PLUS_IO2
+
+#define LL_COMP_INPUT_MINUS_IN0 LL_COMP_INPUT_MINUS_IO1
+#define LL_COMP_INPUT_MINUS_IN1 LL_COMP_INPUT_MINUS_IO2
```
Not a fan of this, c.f. comments in binding files.

erwango replied:
Agreed. You're not supposed to redefine LL symbols (there may be occurrences in tree, but we're working at cleaning them up).

mathieuchopstm commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,378 @@
+	const struct stm32_pclken *pclken;
+	const struct pinctrl_dev_config *pincfg;
+	void (*irq_init)(void);
+	const uint32_t irq_nr;
```
Not sure we need a u32 for this

mathieuchopstm commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,378 @@
+	const struct stm32_pclken *pclken;
+	const struct pinctrl_dev_config *pincfg;
+	void (*irq_init)(void);
+	const uint32_t irq_nr;
+	const LL_COMP_InitTypeDef comp_config;
+	const uint32_t exti_line;
+	const uint8_t exti_line_number;
+	const bool lock_enable;
+	const bool miller_effect_hold_enable;
```
`const` qualifiers are unnecessary here

(FTR: the `const` in `const struct stm32_pclken *` applied to the pointed-to type, not the member, and thus should not be removed)

mathieuchopstm reviewed (changes requested):
Lots of tiny things here and there, but seems overall OK.

Splitting all the EXTI stuff to a separate helper header would improve readability (hopefully, we can merge rework before this, which should solve this issue entirely).

KozhinovAlexander commented:
> Lots of tiny things here and there, but seems overall OK.
> 
> Splitting all the EXTI stuff to a separate helper header would improve readability (hopefully, we can merge rework before this, which should solve this issue entirely).
> 

Thank you for a great review. I agree with all the remarks. It would take me few days, till I resolve the issues. Please wait.

erwango commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,378 @@
+	irq_disable(cfg->irq_nr);
+	LL_COMP_Disable(cfg->comp);
+
+	ret = (int)LL_EXTI_Init(&exti_init_data);
```
We need to get read of all usages of `LL_FOO_Init()` functions are they're not future proof (they won't be available in next STM32 series). Not the best time to introduce a new one.
Please rework the function so it is not used.

(Though, I guess this shouldn't be required following EXTI PR merge)

erwango commented on drivers/comparator/comparator_stm32_comp.c:
```diff
@@ -0,0 +1,378 @@
+		return ret;
+	}
+
+	ret = LL_COMP_Init(cfg->comp, (LL_COMP_InitTypeDef *)&cfg->comp_config);
```
Ditto on the `LL_FOO_Init` usage. Please rework the function to avoid using it.

KozhinovAlexander commented:
@erwango I've already prepared comparator based on new EXTI implementation. Thus please wait till EXTI PR is merged.

etienne-lms reviewed (commented):
@KozhinovAlexander, could you address or reply to @mathieuchopstm review comments?

KozhinovAlexander commented:
@etienne-lms Hi and thank you for your review. I think, I shall put this one PR into draft state, since it will be changed after this one #85508 is merged. This PR will be much simpler and more easier to review then. If you wish to test #85508 on your HW - you're wellcome :)

KozhinovAlexander commented:
This PR contains a lot of changes, which were done before EXTI was merged.
With merged EXTI driver it makes more sense to reopen this PR, s.t. it will stay much more simpler for review.
